hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -479,6 +510,200 @@ public Hashtable<String, String> getPathStatus(final Path path,
     }
   }
 
+  /**
+   * Orchestrates the copying of blob from given source to a given destination.
+   * @param srcPath source path
+   * @param dstPath destination path
+   * @param tracingContext object of TracingContext used for the tracing of the
+   * server calls.
+   * @throws AzureBlobFileSystemException exception thrown from the server calls,
+   * or if it is discovered that the copying is failed or aborted.
+   */
+  @org.apache.hadoop.classification.VisibleForTesting
+  void copyBlob(Path srcPath,
+      Path dstPath,
+      TracingContext tracingContext) throws AzureBlobFileSystemException {
+    AbfsRestOperation copyOp = null;
+    try {
+      copyOp = client.copyBlob(srcPath, dstPath,
+          tracingContext);
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        final BlobProperty dstBlobProperty = getBlobProperty(dstPath,
+            tracingContext);
+        try {
+          if (dstBlobProperty.getCopySourceUrl() != null &&
+              ("/" + client.getFileSystem() + srcPath.toUri().getPath()).equals(
+                  new URL(dstBlobProperty.getCopySourceUrl()).toURI()
+                      .getPath())) {
+            return;
+          }
+        } catch (URISyntaxException | MalformedURLException e) {
+          throw new RuntimeException(e);
+        }
+      }
+      throw ex;
+    }
+    final String progress = getCopyBlobProgress(copyOp);
+    if (COPY_STATUS_SUCCESS.equals(progress)) {
+      return;
+    }
+    final String copyId = copyOp.getResult().getResponseHeader(X_MS_COPY_ID);
```

Comment:
Check for null and hasResult, can result in null pointer exception

Reply:
hasResult is required only in the case when recovering from exception thrown from abfsHttpOperation, abfsRestOperation. Here we are not in the catch block. AbfsRestOperation will always have result, as other wise there will be an exception thrown wherein processResponse is not completed, for which code will never reach this line. it will return back from 
`copyOp = client.copyBlob(srcPath, dstPath,
          tracingContext);`